ynr/apps/sopn_parsing/helpers/pdf_helpers.py

```diff
@@ -22,46 +25,79 @@
 class SOPNDocument:
     def __init__(self, file, all_documents_with_source):
         self.file = file
-        self.pages = []
-        self.all_documents_with_source = all_documents_with_source
-        self.parse_pages()
-        self.matched_pages = {}
-        self.unmatched_documents = list(self.all_documents_with_source)
+        self.unmatched_documents = set(doc for doc in all_documents_with_source)
+        self.matched_documents = set()
+        self.pages = self.parse_pages()
         self.document_heading = self.pages[0].get_page_heading_set()
 
         if len(self.document_heading) < 10:
             raise NoTextInDocumentError()
 
-    def match_all_page(self):
-        """
-        return [{SOPNDocument: str_of_page_numbers }]
-        """
+    @property
+    def matched_pages(self):
+        return [p for p in self.pages if p.matched]
 
-        if len(self.pages) == 1:
-            return [self.unmatched_documents[0], "all"]
-
-        last_page = None
-        for page in self.pages:
-            # get ward sooner to use in detect_top_page
-            print(page.page_number)
-            ward = None
-            if self.unmatched_documents:
-                doc = self.match_page_to_document(page)
-                if doc:
-                    ward = doc.ballot.post.label
+    @property
+    def unmatched_pages(self):
+        return [p for p in self.pages if not p.matched]
 
-            assert any([ward, last_page])
+    def match_ballot_to_pages(self, ballot: Ballot) -> str:
+        """
+        Given a ballot object, loop over all the unmatched pages looking for the
+        first match.
 
-            if page.detect_top_page(self.document_heading, ward, last_page):
-                self.matched_pages.update({page.page_number: page})
-                top_page = page
-            last_page = page
+        Return a string in the format of:
+           "1,2,3,4"
+        OR
+            "all"
 
-        for doc, page_numbers in self.matched_pages.items():
-            page_numbers = ",".join(self.pages_for_ballot)
-            yield [doc, page_numbers]
+        For more on this format, see https://camelot-py.readthedocs.io/en/master/user/quickstart.html?highlight=pages#specify-page-numbers
+        """
+        page_numbers = []
+        previous_page = None
+        for page in self.unmatched_pages:
+            poist_label = ballot.post.label
+            if previous_page:
+                if page.is_continuation_page(
+                    self.document_heading, previous_page, poist_label
+                ):
+                    page_numbers.append(page.page_number)
+                else:
+                    break
+            if page.matches_ward_name(poist_label):
+                page_numbers.append(page.page_number)
+                page.matched = True
+                previous_page = page
+        return ",".join([str(page_num) for page_num in page_numbers])
```

Comment:
Thinking about this, it would be good to assert that the pages are consecutive in the list. If they're not, I think we should raise a `ValueError`.

You can check this in a new method on this class, with logic a bit like:
```py
pages = [1,2,3,4,5,13]

lower = min(pages)
upper = max(pages) + 1

assert pages == range(lower, upper)
```